### PR TITLE
Rewards cap tests

### PR DESCRIPTION
**Pull request by dimpar**
Depends on https://github.com/keep-network/keep-ecdsa/pull/584
Refs #446

Adding tests around the maximum cap rewards per ECDSA keep operator.

### CODE REVIEW

File: solidity/test/rewards/TestECDSARewards.js [hunk 1]
```diff
@@ -230,10 +230,130 @@ describe("ECDSARewards", () => {
...
+
+      for (let i = 0; i < operators.length; i++) {
+        await expectRevert(
+          rewardsContract.withdrawRewards(0, operators[i], {
+            from: operators[i],
+          }),
+          "No rewards to withdraw"
+        )
+      }
+    })
```
Comment (pdyraga) [inline solidity/test/rewards/TestECDSARewards.js:264]: It's worth adding yet one test that will cover `receiveReward + withdrawRewards + receiveRewards + withdrawRewards` path for the same interval.

File: solidity/test/rewards/TestECDSARewards.js [hunk 2]
```diff
@@ -230,10 +230,130 @@ describe("ECDSARewards", () => {
...
+
+      for (let i = 0; i < operators.length; i++) {
+        await rewardsContract.withdrawRewards(0, operators[i], {
+          from: operators[i],
+        })
+      }
+
+      await assertAllocatedRewards(expectedBalanceFromTwoKeeps, 0, operators)
+      await assertWithdrawableRewards(new BN(0), 0, operators)
+      await assertWithdrawnRewards(expectedBalanceFromTwoKeeps, 0, operators)
```
Comment (pdyraga) [inline solidity/test/rewards/TestECDSARewards.js:308]: I think it's time to reorganize tests a little and separate allocation from withdrawal:
```
    rewards allocation
      should not be possible when a keep is not closed 
      should not be possible when a keep is terminated
      should exclude terminated groups
      should cap the maximum reward per beneficiary
    rewards withdrawal
      should correctly distribute rewards between beneficiaries
      should revert when rewards have been withdrawn already 
      should correctly receive rewards from multiple keeps
      should correctly track withdrawable and withdrawn amount
```

File: solidity/test/rewards/TestECDSARewards.js (hunk 1 above)
Comment (dimpar) [reply solidity/test/rewards/TestECDSARewards.js:291]: I've added for: `receiveReward + withdrawRewards + receiveRewards` since the second `receiveRewards` will trigger revert: `Rewards already claimed`

File: solidity/test/rewards/TestECDSARewards.js (hunk 2 above)
Comment (dimpar) [reply solidity/test/rewards/TestECDSARewards.js]: Yes, agree. Reorganized :)

File: solidity/contracts/test/ECDSARewardsStub.sol [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+            _tokenStakingAddress
+        )
+    {
+        beneficiaryRewardCap = 10000 * 10**18;
+    }
+
+    function getBeneficiaryRewardCap() public view returns (uint256) {
+        return beneficiaryRewardCap;
+    }
+
```
[the remark this answers is not in the document]
Comment (dimpar) [reply solidity/contracts/test/ECDSARewardsStub.sol]: Right, removed

File: solidity/test/rewards/TestECDSARewards.js (hunk 1 above)
Comment (pdyraga) [reply solidity/test/rewards/TestECDSARewards.js:291]: That's very good but what I suggested is yet another scenario. The scenario I have in mind is when you claim even more rewards by executing the second `receiveReward` for another keep. Real-world example is when that second keep gets closed later, once you have already received and withdrawn reward for the first keep.

Comment (dimpar) [reply solidity/test/rewards/TestECDSARewards.js:291]: Okay, added in 09453fd

File: solidity/contracts/test/ECDSARewardsStub.sol [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    )
+        public
+        ECDSARewards(
+            _token,
+            _factoryAddress,
+            _tokenStakingAddress
+        )
+    {
+        beneficiaryRewardCap = 10000 * 10**18;
+    }
```
Comment (pdyraga) [inline solidity/contracts/test/ECDSARewardsStub.sol:21]: Wouldn't it be cleaner to provide a setter on the stub that would be called with a value provided by test?

Comment (dimpar) [reply solidity/contracts/test/ECDSARewardsStub.sol:21]: Hm, I don't know if it is cleaner, because it is a "constant" value that should probably not be changed in tests just like in a real `ECDSARewards` contract.  I believe we have a similar example in `TokenStakingStub` where we declare `minimumStake` without a setter. 
But if you think we should have a setter, I can certainly add it.

File: solidity/test/rewards/TestECDSARewards.js [hunk 5]
```diff
@@ -268,11 +342,88 @@ describe("ECDSARewards", () => {
...
+      await assertWithdrawnRewards(new BN(0), 0, operators)
+
+      await withdrawRewards(0, operators)
+
+      await assertWithdrawableRewards(new BN(0), 0, operators)
+      await assertWithdrawnRewards(expectedBalanceFromTwoKeeps, 0, operators)
+    })
   })
 
   async function assertKeepBalanceOfBeneficiaries(expectedBalance) {
```
Comment (pdyraga) [inline solidity/test/rewards/TestECDSARewards.js:424]: Can we add `operators` as a parameter here? Just to stay consistent with the rest of `assert*` functions.

File: solidity/contracts/test/ECDSARewardsStub.sol (hunk 4 above)
Comment (pdyraga) [reply solidity/contracts/test/ECDSARewardsStub.sol:21]: I mean... we update this value for tests and we have a good reason for doing that. I think we should either have a setter and set the value on a stub from tests or don't have a stub at all and test against the real value. Latter is not possible with the test harness we have so I'd go with option one.

**Comment by pdyraga** [comment]
Two last comments and I am good here.

File: solidity/test/rewards/TestECDSARewards.js (hunk 5 above)
Comment (dimpar) [reply solidity/test/rewards/TestECDSARewards.js:424]: Hm, `operators` are actually visible by all the functions in tests which is set in `before`. Maybe let's do the other way around and remove this parameter from `assert*` functions?

Comment (pdyraga) [reply solidity/test/rewards/TestECDSARewards.js:424]: Yeah, we can do that.

File: solidity/contracts/test/ECDSARewardsStub.sol (hunk 4 above)
Comment (dimpar) [reply solidity/contracts/test/ECDSARewardsStub.sol:26]: Done

**Review by pdyraga — APPROVED** [review]
